crates/archive/src/sink.rs

```diff
@@ -0,0 +1,156 @@
+use crate::ingest::ingest_from_service;
+use crate::layout::ChunkWriter;
+use crate::metrics;
+use crate::processor::LineProcessor;
+use crate::progress::Progress;
+use crate::writer::WriterItem;
+use futures_util::TryStreamExt;
+use sqd_data_core::BlockNumber;
+use std::num::NonZeroUsize;
+use std::pin::pin;
+use std::time::{Duration, Instant};
+use tokio::sync::mpsc::UnboundedSender;
+use url::Url;
+
+
+pub struct Sink {
+    processor: LineProcessor,
+    chunk_writer: ChunkWriter,
+    chunk_size: usize,
+    max_num_rows: usize,
+    progress: Progress,
+    url: Url,
+    last_block: Option<BlockNumber>,
+    chunk_sender: UnboundedSender<WriterItem>,
+}
+
+
+impl Sink {
+    pub fn new(
+        processor: LineProcessor,
+        chunk_writer: ChunkWriter,
+        chunk_size: usize,
+        url: Url,
+        last_block: Option<BlockNumber>,
+        chunk_sender: UnboundedSender<WriterItem>,
+    ) -> Sink {
+        let window_size = NonZeroUsize::new(10).unwrap();
+        let granularity = Duration::from_secs(1);
+        let progress = Progress::new(window_size, granularity);
+        Sink {
+            processor,
+            chunk_writer,
+            chunk_size,
+            max_num_rows: 200_000,
+            chunk_sender,
+            progress,
+            url,
+            last_block,
+        }
+    }
+
+    pub async fn r#loop(&mut self) -> anyhow::Result<()> {
+        let mut chunk_first_block = self.chunk_writer.next_block();
+        let mut next_block = self.chunk_writer.next_block();
+        let mut last_report = Instant::now();
+
+        'outer: loop {
+            let mut stream = pin!(ingest_from_service(
+                self.url.clone(),
+                next_block,
+                self.last_block
+            ));
+            let mut data_ingested = false;
+
+            loop {
+                let line = match stream.try_next().await {
+                    Ok(Some(line)) => line,
+                    Ok(None) => {
+                        if let Some(last_block) = self.last_block {
+                            if last_block == self.processor.last_block() {
+                                break 'outer;
+                            }
+                        }
+                        if !data_ingested {
+                            tracing::info!("no blocks were found. waiting 5 min for a new try");
+                            tokio::time::sleep(Duration::from_secs(300)).await;
```

Comment:
Shouldn't this delay be configurable? Sounds like 5 minutes may be a lot for some networks